cpp/src/util/main/UnitTests.cpp

```diff
@@ -125,6 +125,343 @@ TEST(eigen_util, sort_columns) {
+  // check mean convergence with M according to Central Limit Theorem
+  // 99% confidence interval
+  Eigen::VectorXf mean = X.colwise().mean();
+  for (int i = 0; i < N; i++) {
+    EXPECT_NEAR(mean(i), 1.0 / N, 2.576 * std::sqrt(expected_var / M));
+  }
+}
+
+TEST(eigen_util, sort_columns_one_element) {
+  constexpr int kNumRows = 1;
+  constexpr int kMaxNumCols = 1;
+  using Array = Eigen::Array<float, kNumRows, Eigen::Dynamic, 0, kNumRows, kMaxNumCols>;
+
+  Array array{
+      {3},
+  };
+
+  array = eigen_util::sort_columns(array);
+
+  Array expected{
+      {3},
+  };
+
+  for (int r = 0; r < array.rows(); r++) {
+    for (int c = 0; c < array.cols(); c++) {
+      EXPECT_EQ(array(r, c), expected(r, c));
+    }
+  }
+}
+
+TEST(eigen_util, softmax_Array) {
+  constexpr int N = 4;
+  using Array = eigen_util::FArray<N>;
+
+  Array array{{0, 1, 2, 3}};
+  Array expected{{0.0320586, 0.0871443, 0.2368828, 0.6439143}};
+
+  array = eigen_util::softmax<Array>(array);
+
+  for (int i = 0; i < N; ++i) {
+    EXPECT_NEAR(array(i), expected(i), 1e-5);
+  }
+}
+
+TEST(eigen_util, softmax_Tensor) {
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<N, 1>>;
+
+  Tensor tensor;
+  tensor.setValues({{0}, {1}, {2}, {3}});
+  Tensor expected;
+  expected.setValues({{0.0320586}, {0.0871443}, {0.2368828}, {0.6439143}});
+
+  tensor = eigen_util::softmax<Tensor>(tensor);
+
+  for (int i = 0; i < N; ++i) {
+    EXPECT_NEAR(tensor(i, 0), expected(i, 0), 1e-5);
+  }
+}
+
+TEST(eigen_util, sigmoid) {
+  constexpr int N = 4;
+  using Array = eigen_util::FArray<N>;
+
+  Array array{{0, 1, 2, 3}};
+  Array expected{{0.5, 0.7310586, 0.8807971, 0.9525741}};
+
+  array = eigen_util::sigmoid<Array>(array);
+
+  for (int i = 0; i < N; ++i) {
+    EXPECT_NEAR(array(i), expected(i), 1e-5);
+  }
+}
+
+TEST(eigen_util, reverse) {
+  constexpr int M = 3;
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<3, 4>>;
+
+  Tensor tensor;
+  tensor.setValues({{0, 1, 2, 3}, {4, 5, 6, 7}, {8, 9, 10, 11}});
+  Tensor expected;
+  expected.setValues({{3, 2, 1, 0}, {7, 6, 5, 4}, {11, 10, 9, 8}});
+
+  Tensor reversed = eigen_util::reverse<Tensor>(tensor, 1);
+
+  for (int i = 0; i < M; ++i) {
+    for (int j = 0; j < N; ++j) {
+      EXPECT_EQ(reversed(i, j), expected(i, j));
+    }
+  }
+}
+
+TEST(eigen_util, normalize) {
+  constexpr int M = 2;
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<M, N>>;
+  using Array = Eigen::Array<float, M, N, Eigen::RowMajor>;
+
+  Array array{{0, 1, 2, 3}, {0, 0, 4, 0}};
+  Array expected = array / array.sum();
+  Tensor tensor = Eigen::TensorMap<Tensor>(array.data(), array.rows(), array.cols());
+
+  bool success = eigen_util::normalize<Tensor>(tensor, 1e-5);
+
+  EXPECT_TRUE(success);
+  for (int i = 0; i < M; ++i) {
+    for (int j = 0; j < N; ++j) {
+      EXPECT_NEAR(tensor(i, j), expected(i, j), 1e-5);
+    }
+  }
+}
+
+TEST(eigen_util, sample) {
+  constexpr int N = 4;
```

Comment:
At the start of this function, can you call,

```
util::Random::set_seed(1);
```

This ensures that the `eigen_util::sample()` calls will be consistent across runs.

As written, it will be consistent by virtue of the `util::Random::set_seed()` call that happens earlier (in the `test_zero_out()` call). But as soon as someone adds another test case that gets called before this one, that also uses `util::Random`, this test case will use different random values, and could end up failing unexpectedly.